planner/core/rule_join_reorder_greedy.go

```diff
@@ -130,3 +138,125 @@ func (s *joinReorderGreedySolver) checkConnectionAndMakeJoin(leftNode, rightNode
 	})
 	return s.newJoinWithEdges(leftNode, rightNode, usedEdges, otherConds), remainOtherConds
 }
+
+// For curCostBasedReorderGreedySolver
+// solve reorders the join nodes in the group based on a greedy algorithm.
+//
+// For each node having a join equal condition with the current join tree in
+// the group, calculate the join cost of that node and the join
+// tree, choose the node with the smallest cost to join with the current join tree.
+//
+// join cost = RowCount(join)
+// TODO: this formula can be changed to real physical cost in future.
+//
+// For the nodes and join trees which don't have a join equal condition to connect them,
+// we make a join tree with greedy strategy to do the cartesian joins finally.
+func (s *curCostBasedReorderGreedySolver) solve(joinNodePlans []LogicalPlan) (LogicalPlan, error) {
+	for _, node := range joinNodePlans {
+		_, err := node.recursiveDeriveStats()
+		if err != nil {
+			return nil, err
+		}
+		s.curJoinGroup = append(s.curJoinGroup, &jrNode{
+			p:       node,
+			cumCost: node.statsInfo().RowCount,
+		})
+	}
+
+	var cartesianGroup []LogicalPlan
+	for len(s.curJoinGroup) > 0 {
+		newNode, err := s.constructConnectedJoinTree()
+		if err != nil {
+			return nil, err
+		}
+		cartesianGroup = append(cartesianGroup, newNode.p)
+	}
+
+	return s.makeHuffJoin(cartesianGroup)
+}
+
+func (s *curCostBasedReorderGreedySolver) constructConnectedJoinTree() (*jrNode, error) {
+	// find the jrNode with smallest cumCost as the root of Join Tree
+	root_ips := 0
+	for i, node := range s.curJoinGroup {
+		if node.cumCost < s.curJoinGroup[root_ips].cumCost {
+			root_ips = i
+		}
+	}
+	curJoinTree := s.curJoinGroup[root_ips]
+	s.curJoinGroup = append(s.curJoinGroup[:root_ips], s.curJoinGroup[root_ips+1:]...)
+
+	for {
+		bestCost := math.MaxFloat64
+		bestIdx := -1
+		var finalRemainOthers []expression.Expression
+		var bestJoin LogicalPlan
+		for i, node := range s.curJoinGroup {
+			newJoin, remainOthers := s.checkConnectionAndMakeJoin(curJoinTree.p, node.p)
+			if newJoin == nil {
+				continue
+			}
+			_, err := newJoin.recursiveDeriveStats()
+			if err != nil {
+				return nil, err
+			}
+			curCost := newJoin.statsInfo().RowCount
+			if bestCost > curCost {
+				bestCost = curCost
+				bestJoin = newJoin
+				bestIdx = i
+				finalRemainOthers = remainOthers
+			}
+		}
+		// If we could not find more join node, meaning that the sub connected graph have been totally explored.
+		if bestJoin == nil {
+			break
+		}
+		curJoinTree = &jrNode{
+			p:       bestJoin,
+			cumCost: bestCost,
+		}
+		s.curJoinGroup = append(s.curJoinGroup[:bestIdx], s.curJoinGroup[bestIdx+1:]...)
+		s.otherConds = finalRemainOthers
+	}
+	return curJoinTree, nil
+}
+
+// upMinCostPlan find the plan with minimum cost in cur[ips:] and swap it to cur[ips]
+func (s *curCostBasedReorderGreedySolver) upMinCostPlan(cur *[]LogicalPlan, ips int) {
```

Comment:
Use https://golang.org/pkg/container/heap/ instead?

Reply:
When implementing I found that using heap will introduce 6 functions and 1 interface, while experting small impact on the speed of planner since the overall implement of join_reorder_greedy is `O(sqr(n))`. So I think that using simply one function is better. @eurekaka